scripts/ASTImportTest.sh

```diff
@@ -153,6 +175,281 @@ function test_ast_import_export_equivalence
     TESTED=$((TESTED + 1))
 }
 
+function split_combined_json
+{
+    local json_file="$1"
+    local output_path="$2"
+    local prefix="${3:-}"
+    for path_with_contract in $(jq '.contracts | keys | .[]' "${json_file}" 2> /dev/null)
+    do
+        local path=${path_with_contract}
+        local contract=""
+        local delimiter
+        delimiter=$("${EXPR}" index "${path}" ":") || true
+        if [[ -z "${prefix}" ]]
+        then
+            path=${path_with_contract:0:((${delimiter} - 1))}
+            contract=${path_with_contract:((${delimiter})):((${#path_with_contract} - ${delimiter} - 1))}
```

Comment:
This will fail when we have some tests with multiple sources, for instance, consider this case: https://github.com/ethereum/solidity/blob/16ae76cad7ca544a183e8340216a4a72d2a08dcd/test/libsolidity/semanticTests/expressions/module_from_ternary_expression.sol#L4

The current code would generate `yul/A_C.yul.bin` and `yul/B_C.yul.bin` but it would expect only `yul/C.yul.bin`. Not really sure how to deal with it though.

Reply:
Please see the proposed fix in this commit: https://github.com/ethereum/solidity/pull/13673/commits/62ff0fb0cbbf724f79b6125dfaabf33f6fcf807c

Reply:
I couldn't manage to properly fix this case, so I workaround it by renaming the contract in source `A` from `C` to something else, e.g. `D`, so the filenames when splitting the `combined.json` remains consistent between the yul and solidity files, e.g. `sol/C.bin` and `yul/C.yul.bin` and only the last element of the path is considered to generate the outputs that will be compared after importing (see: function [split_combined_json](https://github.com/ethereum/solidity/pull/13673/files#diff-29a6c42ce3ccf40f99f34c261d5d39d6b5e1103696bb7859957445a203e3b17aR190) in `ASTImportTest.sh`).

The root of the problem seems to be this function and the fact that the compiler outputs different filenames per output component.

For instance, consider these two test cases:
```
1.
==== Source: A ====
contract C {}

==== Source: B ====
import "A" as M;
contract C {}
```

```
2.
==== Source: D ====
contract C {}

==== Source: E ====
import "D" as N;
contract E {}
```

The only difference between them is that in the first contract `C` is defined in both sources `A` and `B`. This changes the filename outputted produced by the compiler.

After splitting the sources for each test case, if we compile multiple files at once, e.g.: compiling `A` and `B` with `solc --bin A B -o sol_AB/`, the CLI outputs: `A_C.bin  B_C.bin`, however compiling `D` and `E` with the same command, i.e.: `solc --bin D E -o sol_DE/`, the CLI outputs: `C.bin  E.bin`. If you compile the 4 contracts, i.e.: `solc --bin A B D E -o sol_ABDE/` the CLI outputs:  `A_C.bin  B_C.bin  D_C.bin  E.bin`.

Compiling each source individually yields the following:

```sh
for contract in A B D E; do solc --bin $contract -o sol_$contract; done ; ls sol_*
sol_A:
C.bin

sol_B:
A_C.bin  B_C.bin

sol_D:
C.bin

sol_E:
C.bin  E.bin
```

And using the `--combined-json` output option, we have:
```sh
for contract in A B D E; do solc --combined-json "bin" $contract -o sol_$contract; done ; jq '.contracts | keys' sol_*/combined.json
 
[
  "A:C"
]
[
  "A:C",
  "B:C"
]
[
  "D:C"
]
[
  "D:C",
  "E:E"
]
```
Maybe the compiler should always include the source path to disambiguate the contracts as is done in the standard-json output. However in some cases the path is a file path to the solidity source, e.g. `<somepath>.sol:C`.

Furthermore, the problem becomes evident when you need to re-import the evm assembly, for instance:

```sh
# This will generate `yul/C.yul`
solc --ir A -o yul
# Then we export the asm. Here, the script could try to infer the correct output name (i.e. `A_C`). However, as shown above
# The output is not always consistent, since it depends on how many files are being compiled at the same time and the structure of the contracts
solc --strict-assembly yul/C.yul --asm-json --pretty-json --json-indent 4 > yul/C.asm
grep -v '^=======' yul/C.asm > yul/C.asm.json
solc --combined-json "bin" --pretty-json --json-indent 4 --import-asm-json yul/C.asm.json
```

The commands above will produce the following `combined.json`:
```
{
    "contracts":
    {
        "yul/C.asm.json":
        {
            "bin": "608060405234601957600e601d565b603e60288239603e90f35b6023565b60405190565b5f80fdfe60806040525f80fdfea26469706673582212205bfd3a9a5810555cc61103cf3fb849e04c003ceeff728bfa10f89eaa5da8c29364736f6c63430008160033"
        }
    },
    "version": "0.8.22+commit.7508c5b1.mod.Linux.g++"
}
```
As you can see, the name of the contract is the filename including the given path when compiling, and maybe it should be `A:C` to be consistent with the output when using only `--combined-json` option.

Reply:
> I couldn't manage to properly fix this case, so I workaround it by renaming the contract in source A from C to something else,

That workaround is the best solution here :) We don't have to handle this in full generality. We're not testing the handling of contracts with conflicting names here so just ensuring that we don't have such contracts is fine.